libfreerdp/core/gateway/websocket.c

```diff
@@ -0,0 +1,497 @@
+	else if (len < 0x10000)
+		fullLen = len + 8; /* 2 byte "mini header" + 2 byte length + 4 byte masking key */
+	else
+		fullLen = len + 14; /* 2 byte "mini header" + 8 byte length + 4 byte masking key */
+
+	sWS = Stream_New(NULL, fullLen);
+	if (!sWS)
+		return FALSE;
+
+	winpr_RAND((BYTE*)&maskingKey, 4);
+
+	Stream_Write_UINT8(sWS, WEBSOCKET_FIN_BIT | opcode);
+	if (len < 126)
+		Stream_Write_UINT8(sWS, len | WEBSOCKET_MASK_BIT);
+	else if (len < 0x10000)
+	{
+		Stream_Write_UINT8(sWS, 126 | WEBSOCKET_MASK_BIT);
+		Stream_Write_UINT16_BE(sWS, len);
+	}
+	else
+	{
+		Stream_Write_UINT8(sWS, 127 | WEBSOCKET_MASK_BIT);
+		Stream_Write_UINT32_BE(sWS, 0); /* payload is limited to INT_MAX */
+		Stream_Write_UINT32_BE(sWS, len);
+	}
+	Stream_Write_UINT32(sWS, maskingKey);
+
+	/* mask as much as possible with 32bit access */
+	for (streamPos = 0; streamPos + 4 <= len; streamPos += 4)
+	{
+		uint32_t data;
+		Stream_Read_UINT32(sPacket, data);
+		Stream_Write_UINT32(sWS, data ^ maskingKey);
+	}
+
+	/* mask the rest byte by byte */
+	for (; streamPos < len; streamPos++)
+	{
+		BYTE data;
+		BYTE* partialMask = ((BYTE*)&maskingKey) + (streamPos % 4);
+		Stream_Read_UINT8(sPacket, data);
+		Stream_Write_UINT8(sWS, data ^ *partialMask);
+	}
+
+	Stream_SealLength(sWS);
+
+	ERR_clear_error();
+	status = BIO_write(bio, Stream_Buffer(sWS), Stream_Length(sWS));
+	Stream_Free(sWS, TRUE);
+
+	if (status != (SSIZE_T)fullLen)
+		return FALSE;
+
+	return TRUE;
+}
+
+int websocket_write(BIO* bio, const BYTE* buf, int isize, WEBSOCKET_OPCODE opcode)
+{
+	size_t payloadSize;
+	size_t fullLen;
+	int status;
+	wStream* sWS;
+
+	uint32_t maskingKey;
+
+	int streamPos;
+
+	winpr_RAND((BYTE*)&maskingKey, 4);
+
+	payloadSize = isize;
+	if ((isize < 0) || (isize > UINT16_MAX))
+		return -1;
+
+	if (payloadSize < 126)
+		fullLen = payloadSize + 6; /* 2 byte "mini header" + 4 byte masking key */
+	else if (payloadSize < 0x10000)
+		fullLen = payloadSize + 8; /* 2 byte "mini header" + 2 byte length + 4 byte masking key */
+	else
+		fullLen = payloadSize + 14; /* 2 byte "mini header" + 8 byte length + 4 byte masking key */
+
+	sWS = Stream_New(NULL, fullLen);
+	if (!sWS)
+		return FALSE;
+
+	Stream_Write_UINT8(sWS, WEBSOCKET_FIN_BIT | opcode);
+	if (payloadSize < 126)
+		Stream_Write_UINT8(sWS, payloadSize | WEBSOCKET_MASK_BIT);
+	else if (payloadSize < 0x10000)
+	{
+		Stream_Write_UINT8(sWS, 126 | WEBSOCKET_MASK_BIT);
+		Stream_Write_UINT16_BE(sWS, payloadSize);
+	}
+	else
+	{
+		Stream_Write_UINT8(sWS, 127 | WEBSOCKET_MASK_BIT);
+		/* biggest packet possible is 0xffff + 0xa, so 32bit is always enough */
+		Stream_Write_UINT32_BE(sWS, 0);
+		Stream_Write_UINT32_BE(sWS, payloadSize);
+	}
+	Stream_Write_UINT32(sWS, maskingKey);
+
+	/* mask as much as possible with 32bit access */
+	for (streamPos = 0; streamPos + 4 <= isize; streamPos += 4)
+	{
+		uint32_t masked = *((const uint32_t*)((const BYTE*)buf + streamPos)) ^ maskingKey;
```

Comment:
not sure this is endian safe? (there are people using it on such systems I´ve been told ;))